spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

```diff
@@ -56,6 +60,107 @@
 		return results.iterator().next();
 	}
 
+	/**
+	 * Return a single result object from the given Stream.
+	 * <p>Returns {@code null} if 0 result objects found;
+	 * throws an exception if more than 1 element found.
+	 * @param results the result Stream (can be {@code null})
+	 * @return the single result object, or {@code null} if none
+	 * @throws IncorrectResultSizeDataAccessException if more than one
+	 * element has been found in the given Stream
+	 */
+	@Nullable
+	public static <T> T singleResult(@Nullable Stream<T> results) throws IncorrectResultSizeDataAccessException {
+		if (results == null) {
+			return null;
+		}
+		List<T> resultList = results.limit(2).toList();
```

Comment:
There is a trade off here and in `#optionalResult` whether we should call `#close` on the stream:

On one hand Java convention is the code who creates a "resource" is responsible for disposing it, in this case the caller of this method would be responsible to call `#close` on the streams where required.
On the other hand failing to call `#close` on some streams, like the ones returned by `JdbcOperations#queryForStream`, can lead to resource leaks. Having the caller put streams in a try-with-resources block is both cumbersome and error prone.

Reply:
Makes sense. The passed stream is now closed in both methods. Also e.g. hibernate's Query.stream() documentation suggests the stream should be closed afterwards, so not so far fetched 👍